### PR TITLE
Filters

### PR DESCRIPTION
**Pull request by amohanan-scottlogic**
# Sort and Filter

- All sort methods and their respective unit tests
- All filters and their respective unit tests

### CODE REVIEW

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]
The filters are looking really good! I've tried not to duplicate my comments, but there will be a few things that apply across all sorters/filters/tests.

In terms of branching strategy - keep the updates to this stuff here on the filters branch, and keep working on your new branch for the combined filters/sorters. I don't think we need the develop branch any more, so we should be able to delete that. We'll probably need to do a bit of rebasing when it comes to merging the branches, but we can go through that together. We can have a chat if any of that doesn't make sense.

File: src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4);
+        List<UserPost> actualResult = sortByFullName.sort(userPosts, SortOrder.ASC);
+        List<UserPost> expectedResult = Arrays.asList(userPost1, userPost2, userPost4, userPost3);
+        Assert.assertEquals(actualResult, expectedResult);
+    }
+
+    @Test
+    public void fullNameSorter_spaceForLastName_correctListReturnedAsc() {
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe ");
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java:40]: Joe and " "

File: src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Test
+    public void fullNameSorter_mixedValidNames_correctListReturnedAsc() {
+        UserPost userPost1 = createUserPost("Joe");
+        UserPost userPost2 = createUserPost("Joe Bloggs");
+        UserPost userPost3 = createUserPost("May Bloggs");
+        UserPost userPost4 = createUserPost("Joe Bloggs");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4);
+        List<UserPost> actualResult = sortByFullName.sort(userPosts, SortOrder.ASC);
+        List<UserPost> expectedResult = Arrays.asList(userPost1, userPost2, userPost4, userPost3);
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java:33]: yes. Just tested it out. It So if there are several Bloggs, the one without a first name comes on the top of the sorted list

File: src/main/java/com/scottlogic/filters/LikesFilter.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class LikesFilter implements PostFilter {
+    @Override
+    public List<UserPost> filter(List<UserPost> inputList) {
+        List<UserPost> filteredList = new ArrayList<>();
+        for (UserPost userPost : inputList) {
+            if (userPost.getLikeCount() >= 1) {
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/filters/LikesFilter.java:13]: >0

[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/filters/LikesFilter.java:13]: I meant to say, check for greater than 0. My '>' disappeared in the comment

File: src/main/java/com/scottlogic/sorts/FullNamePostSorter.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return remainingPosts;
+    }
+
+    private String[] nameSplit(String name) {
+
+        String[] splitted = name.split(" ");
+        String[] names = new String[2];
+        int numberOfNames = splitted.length;
+        names[0] = splitted[0];
+        names[1] = numberOfNames > 1 ? splitted[numberOfNames - 1] : " ";
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/sorts/FullNamePostSorter.java:71]: I'm doing this so that all names with no last name are on the top of the list(when sorting in ascending order), and then they are sorted according to first name.

File: src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java:40]: As discussed, it's actually just "Joe".

File: src/test/java/com/scottlogic/sorts/AuthorPostSorterTest.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Test
+    public void sort_blankNamesAsc_returnBlankFirst() {
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe Blogg");
+        UserPost userPost3 = createUserPost("");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> actualResult = sortByAuthor.sort(userPosts, SortOrder.ASC);
+        List<UserPost> expectedResult = Arrays.asList( userPost3, userPost1, userPost2);
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/sorts/AuthorPostSorterTest.java:64]: Yes it is the other way around. No idea how I've written it this way and got it passed. Have re-written and re-tested :)

File: src/main/java/com/scottlogic/sorts/FullNamePostSorter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply src/main/java/com/scottlogic/sorts/FullNamePostSorter.java:71]: I guess this is a bit of a design choice, but have you thought about what happens if someone just enters their surname?

**Review by pling-scottlogic — COMMENTED** [review]
Still a few test cases to cover, I think. Null posts, null lists and empty lists in particular.

File: src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost(" ");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2);
+        List<UserPost> actualResult = sortByFullName.sort(userPosts, SortOrder.ASC);
+        List<UserPost> expectedResult = Arrays.asList(userPost2, userPost1);
+        Assert.assertEquals(actualResult, expectedResult);
+    }
+
+    @Test
+    public void fullNameSorter_middleNameAsc_ignoreMiddleNameReturned() {
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/sorts/FullNamePostSorterTest.java:67]: Is the middle name completely ignored, or is it treated as an extension of the first name? It would be good to test an input containing something like "Joe A Bloggs" and "Joe B Bloggs".

File: src/test/java/com/scottlogic/filters/DatesFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Test
+    public void filter_borderDatesMax_borderDatesIncludedReturned() {
+        UserPost userPost1 = createUserPost(OffsetDateTime.of(20, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost2 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost3 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 1, ZoneOffset.UTC));
+        UserPost userPost4 = createUserPost(OffsetDateTime.of(2012, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost5 = createUserPost(OffsetDateTime.of(2020, 12, 31, 7, 12, 3, 0, ZoneOffset.UTC));
+
+        DatesFilter filterByDates = new DatesFilter(OffsetDateTime.of(2020, 1, 1, 7, 12, 3, 0, ZoneOffset.UTC), OffsetDateTime.of(2020, 12, 31, 7, 12, 3, 0, ZoneOffset.UTC));
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/DatesFilterTest.java:95]: We can often split code over multiple lines if if gets too long, e.g.
```
DatesFilter filterByDates = new DatesFilter(
    OffsetDateTime.of(2020, 1, 1, 7, 12, 3, 0, ZoneOffset.UTC),
    OffsetDateTime.of(2020, 12, 31, 7, 12, 3, 0, ZoneOffset.UTC));
```
Just be careful with your indentation.

File: src/test/java/com/scottlogic/filters/DatesFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        List<UserPost> expectedResult = Arrays.asList(userPost2, userPost3, userPost5);
+        Assert.assertEquals(actualResult, expectedResult);
+    }
+
+    @Test
+    public void filter_borderDatesMin_borderDatesIncludedReturnedSecondsIgnored() {
+        UserPost userPost1 = createUserPost(OffsetDateTime.of(20, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost2 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 4, 0, ZoneOffset.UTC));
+        UserPost userPost3 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 1, ZoneOffset.UTC));
+        UserPost userPost4 = createUserPost(OffsetDateTime.of(2012, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/DatesFilterTest.java:107]: It would be a lot easier to see what's going on if we named these posts a bit more descriptively. It would also be nice to extract the repeated date into a variable, then use it for setting up both the post and the filter.

**Review by pling-scottlogic — COMMENTED** [review]
Getting there - just a few more things to tidy up. Make sure every one of your tests covers the fringe cases we've spoken about (null fields in post, null post, null list, empty list), as well as ensuring the input isn't mutated.

Also check for any unresolved conversations. I've tried to resolve everything that's been addressed, so look at the rest for any outstanding actions.

File: src/main/java/com/scottlogic/sorts/AuthorPostSorter.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        if (inputList == null) {
+            return listToSort;
+        }
+        listToSort = NullPostChecker.nullPostCheck(inputList);
+
+        switch (orderIn) {
+
+            case ASC -> Collections.sort(listToSort,comparing(UserPost::getAuthor, nullsLast(naturalOrder())));
+
+            case DESC -> Collections.sort(listToSort, Collections.reverseOrder(comparing(UserPost::getAuthor, nullsFirst(naturalOrder()))));
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/sorts/AuthorPostSorter.java:28]: Nice bit of tidy-up here. We'll get it even tidier when we move on to looking at streams!

File: src/test/java/com/scottlogic/filters/ContentFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    public void filter_emptyList_emptyListReturned() {
+        ContentFilter filterByKeyword = new ContentFilter("post");
+        List<UserPost> userPosts = new ArrayList<>();
+        List<UserPost> actualResult = filterByKeyword.filter(userPosts);
+        Assert.assertTrue(actualResult.isEmpty());
+    }
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/ContentFilterTest.java:77]: We're missing a test for a null input list here. It would also be good to see one where the search term is not in the input list.

File: src/test/java/com/scottlogic/filters/DatesFilterTest.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        List<UserPost> expectedResult = Arrays.asList(userPost2, userPost3);
+        Assert.assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    public void filter_nullDateRequested_emptyListReturned() {
+        UserPost userPost1 = createUserPost(OffsetDateTime.of(20, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost2 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+
+        DatesFilter filterByDates = new DatesFilter(OffsetDateTime.of(2020, 1, 1, 7, 12, 3, 0, ZoneOffset.UTC), null);
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/DatesFilterTest.java:44]: Can we have another test where the filter's start date is null?

File: src/test/java/com/scottlogic/filters/DatesFilterTest.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                OffsetDateTime.of(2020, 12, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+
+        UserPost userPost1 = createUserPost(OffsetDateTime.of(20, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+        UserPost userPost2 = createUserPost(OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC));
+
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2);
+        List<UserPost> userPostsCopy = Arrays.asList(userPost1, userPost2);
+        filterByDates.filter(userPosts);
+        Assert.assertEquals(userPostsCopy, userPosts);
+    }
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/DatesFilterTest.java:158]: I think we're missing a test where there's a null date in one of the posts.

File: src/test/java/com/scottlogic/filters/LikesFilterTest.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    public void filter_likesCountWhenPresent_correctListReturned() {
+        LikesFilter filterByLikes = new LikesFilter();
+        UserPost userPost1 = createUserPost(2);
+        UserPost userPost2 = createUserPost(0);
+        UserPost userPost3 = createUserPost(1);
+        UserPost userPost4 = createUserPost(3);
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4);
+        List<UserPost> actualResult = filterByLikes.filter(userPosts);
+        List<UserPost> expectedResult = Arrays.asList(userPost1, userPost3, userPost4);
+        Assert.assertEquals(actualResult, expectedResult);
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/filters/LikesFilterTest.java:31]: Looks like you missed this file when swapping round the actual/expected.

File: src/test/java/com/scottlogic/sorts/AuthorPostSorterTest.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    public void sort_blankNamesDesc_returnBlankLast() {
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe Blogg");
+        UserPost userPost3 = createUserPost("");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> actualResult = sortByAuthor.sort(userPosts, SortOrder.DESC);
+        List<UserPost> expectedResult = Arrays.asList(userPost1, userPost2, userPost3);
+        Assert.assertEquals(actualResult, expectedResult);
+    }
+}
```
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply src/test/java/com/scottlogic/sorts/AuthorPostSorterTest.java:172]: Getting there with these. Try and write something to cover that last bullet point.

File: src/test/java/com/scottlogic/sorts/ContentPostSorterTest.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        List<UserPost> expectedResult = Arrays.asList(userPost1, userPost3, userPost2);
+        Assert.assertEquals(actualResult, expectedResult);
+    }
+
+    @Test
+    public void sort_newLineSamePostsAsc_correctListReturned() {
+        UserPost userPost1 = createUserPost("An example of a post \nwith lines breaks.");
+        UserPost userPost2 = createUserPost("An example of a post with lines breaks.");
+        UserPost userPost3 = createUserPost("Hello World!");
+        UserPost userPost4 = createUserPost("An example of a post without line breaks");
```
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply src/test/java/com/scottlogic/sorts/ContentPostSorterTest.java:53]: Don't forget this!

File: src/test/java/com/scottlogic/filters/DatesFilterTest.java (hunk 12 above)
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/filters/DatesFilterTest.java:158]: I'm not able to test that because null date is not allowed by the Object

**Review by samgladstone — APPROVED** [review]
Good Stuff
